packages/whenable/src/whenable.js

```diff
@@ -0,0 +1,131 @@
+// @ts-check
+import { makePromiseKit } from '@endo/promise-kit';
+import { M } from '@endo/patterns';
+import { makeTagged } from '@endo/pass-style';
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareWhenableKits = zone => {
+  /** WeakMap<object, any> */
+  const whenable0ToEphemeral = new WeakMap();
+
+  /**
+   * Get the current incarnation's promise kit associated with a whenable0.
+   *
+   * @param {import('./types.js').WhenablePayload['whenable0']} whenable0
+   * @returns {import('@endo/promise-kit').PromiseKit<any>}
+   */
+  const findCurrentKit = whenable0 => {
+    let pk = whenable0ToEphemeral.get(whenable0);
+    if (pk) {
+      return pk;
+    }
+
+    pk = makePromiseKit();
+    pk.promise.catch(() => {}); // silence unhandled rejection
+    whenable0ToEphemeral.set(whenable0, pk);
+    return pk;
+  };
+
+  /**
+   * @param {'resolve' | 'reject'} kind
+   * @param {import('./types.js').WhenablePayload['whenable0']} whenable0
+   * @param {unknown} value
+   */
+  const settle = (kind, whenable0, value) => {
+    const kit = findCurrentKit(whenable0);
+    const cb = kit[kind];
+    if (!cb) {
+      return;
+    }
+    whenable0ToEphemeral.set(whenable0, harden({ promise: kit.promise }));
+    cb(value);
+  };
+
+  const rawMakeWhenableKit = zone.exoClassKit(
+    'Whenable0Kit',
+    {
+      whenable0: M.interface('Whenable0', {
+        shorten: M.call().returns(M.promise()),
+      }),
+      settler: M.interface('Settler', {
+        resolve: M.call().optional(M.any()).returns(),
+        reject: M.call().optional(M.any()).returns(),
+      }),
+    },
+    () => ({}),
+    {
+      whenable0: {
+        /**
+         * @returns {Promise<any>}
+         */
+        shorten() {
```

Comment:
Aha! At least here, `shorten` is a method of the `whenable0` facet of an exoClassKit. I see that the Tagged structure is placed around it below.

Reply:
Resolving